openstack/clientconfig/requests.go

```diff
@@ -819,6 +819,7 @@ func NewServiceClient(service string, opts *ClientOpts) (*gophercloud.ServiceCli
 		// Otherwise create a new HTTP client with the generated TLS config.
 		pClient.HTTPClient = http.Client{
 			Transport: &http.Transport{
+				Proxy:           http.ProxyFromEnvironment,
```

Comment:
I wondered why we are doing this in an else statement. Are there other http client options which can be passed in which we're ignoring?

Would it work to unconditionally setting pClient.Transport.TLSClientConfig = tlsConfig?

Reply:
The conditional is for when the user comes with his own http client, in which case this it what we need to use.

The else branch here is really the normal flow.

Reply:
What previously set ProxyFromEnvironment?

Reply:
I assume the default transport (when you don't specify one) takes care of it.

Reply:
> The conditional is for when the user comes with his own http client, in which case this it what we need to use.

Exactly correct. It's impossible to support an HTTPClient that will suit everyone's needs, so having the ability to specify a custom one is important.

With this in mind, I'm OK with this PR. `ProxyFromEnvironment` makes sense and helps read in a few environment variables. The reason it wasn't accounted for before was because this HTTPClient logic was taken from Terraform where `ProxyFromEnvironment` isn't used. I understand how adding this statement in a prior PR removed functionality in other applications, though.

However, if there are additional changes to this area of code in the future, I might pass on merging them. My concern would be that the HTTPClient being built here is beginning to get to "app-specific" territory and the custom client might be a better option.

Reply:
Answering my own question, this was DefaultTransport, which is defined here: https://golang.org/pkg/net/http/

```
var DefaultTransport RoundTripper = &Transport{
    Proxy: ProxyFromEnvironment,
    DialContext: (&net.Dialer{
        Timeout:   30 * time.Second,
        KeepAlive: 30 * time.Second,
    }).DialContext,
    ForceAttemptHTTP2:     true,
    MaxIdleConns:          100,
    IdleConnTimeout:       90 * time.Second,
    TLSHandshakeTimeout:   10 * time.Second,
    ExpectContinueTimeout: 1 * time.Second,
}
```

The implicit behaviour change here goes beyond no longer using the proxy: it looks like we also changed all the various timeouts from their previous values.

Would it be worth, for example, copying DefaultTransport and simply setting Transport.TLSClientConfig in the copy?

Reply:
Thank you for looking into this. I agree that the best method would be to copy the defaults.

Reply:
I'm now making a copy of the DefaultTransport via the Clone() method introduced in golang 1.13.
Are we OK with making golang 1.13 the minimum version?

Otherwise we'll have to copy the DefaultTransport struct here.

Reply:
This was my local patch:

```go
// Otherwise initialise a new HTTP client based on
// DefaultTransport, and add the generated TLS config.
tlsTransport := *(http.DefaultTransport.(*http.Transport))
tlsTransport.TLSClientConfig = tlsConfig
pClient.HTTPClient.Transport = &tlsTransport
```

Does that do what I think it does (namely a shallow copy)?